sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala

```diff
@@ -367,6 +367,22 @@ class TungstenAggregationIterator(
     }
   }
 
+  TaskContext.get().addTaskCompletionListener(_ => {
+    // At the end of the task, update the task's peak memory usage. Since we destroy
+    // the map to create the sorter, their memory usages should not overlap, so it is safe
+    // to just use the max of the two.
+    val mapMemory = hashMap.getPeakMemoryUsedBytes
+    val sorterMemory = Option(externalSorter).map(_.getPeakMemoryUsedBytes).getOrElse(0L)
+    val maxMemory = Math.max(mapMemory, sorterMemory)
+    val metrics = TaskContext.get().taskMetrics()
+    peakMemory += maxMemory
+    spillSize += metrics.memoryBytesSpilled - spillSizeBefore
+    metrics.incPeakExecutionMemory(maxMemory)
```

Comment:
hmm, the description of `peakExecutionMemory` in `TaskMetrics` is:

    ...The value of this accumulator should be approximately the sum of the peak sizes across all such data structures created in this task...

So it is designed to get the sum of peak memory of operators in the task. I think because the operators are not ran in sequence but in an iterator way, it's reasonable to sum the peak memory. Although the peak points of the operators might not at the same moment.